nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,398 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from collections import OrderedDict
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from nni.retiarii.nn.pytorch import LayerChoice, InputChoice
+from .base_lightning import BaseOneShotLightningModule
+from .darts import DartsInputChoice, DartsLayerChoice
+
+
+class DartsModule(BaseOneShotLightningModule):
+    """
+    The DARTS model. In each iteration, there are 2 training phases. The phase 1 is architecture
+    step, in which the model parameters are frozen and the alphas are optimized. The phase 2 is model
+    step, in wchich the alphas are frozen and the model parameters are optimized. In this step, the
+    out put of each choice are sumed up with respect to their alpha value. The result of DARTS
+    is argmax in alpha.
+    The DARTS Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
+
+    Parameters
+    ----------
+    base_model : pl.LightningModule
+        The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+        is wrapped by base_model, and base_model will be wrapped by this model.
+    custom_replace_dict : Dict[Type[nn.Module], Callable[[nn.Module], nn.Module]], default = None
+        The custom xxxChoice replace method. Keys should be xxxChoice type and values should
+        return an nn.module. This custom replace dict will override the default replace
+        dict of each NAS method.
+
+    Reference
+    ----------
+    .. [darts] H. Liu, K. Simonyan, and Y. Yang, “DARTS: Differentiable Architecture Search,” presented at the
+        International Conference on Learning Representations, Sep. 2018. Available: https://openreview.net/forum?id=S1eYHoC5FX
+    """
+    automatic_optimization = False
+
+    def training_step(self, batch, batch_idx):
+        # grad manually, only 1 architecture optimizer for darts
+        opts = self.optimizers()
+        if isinstance(opts,list):
+            # list of optimizers
+            # pylint: disable=unsubscriptable-object
+            arc_optim = opts[0]
+            w_optim = opts[1:]
+        else :
+            # no model optimizer
+            arc_optim = opts
+
+        # ParallelTrainValDataLoader will yield both train and val data in a batch
+        trn_batch, val_batch = batch
+
+        # phase 1: architecture step
+        # The resample_architecture hook is kept for some following darts-based NAS
+        # methods such as proxyless. See code of those methods for details.
+        self.resample_architecture()
+        arc_optim.zero_grad()
+        arc_step_loss = self._extract_user_loss(val_batch, 2 * batch_idx)
```

Comment:
What happens to other metrics returned by `training_step`?

The name of `_extract_user_loss` is very confusing?

Reply:
other metrics are just dropped.

Reply:
As we discussed last time, the other metrics should be returned. They can't be simply dropped because they could be used by users later in `training_epoch_end`.